File: pyBMD_modules/pre_pyBMD_normalize_filter.py
import sys
from pathlib import Path

import numpy as np
import pandas as pd
from pydeseq2.dds import DeseqDataSet
from pydeseq2.ds import DeseqStats


def parse_args():
    if len(sys.argv) != 7:
        print("Usage: python normalize_and_filter.py <count_file> <annotation_file> "
              "<cell_line> <compound_name> <control_names> <output_dir>")
        sys.exit(1)

    return {
        "count_file": sys.argv[1],
        "annotation_file": sys.argv[2],
        "cell_line": sys.argv[3],
        "compound_name": sys.argv[4],
        "control_names": [c.strip() for c in sys.argv[5].split(",")],
        "output_dir": sys.argv[6],
    }


def load_and_subset(count_file, annotation_file, cell_line, compound_name, control_names):
    """Load data and subset to relevant samples (cell line + compound + controls)."""
    counts = pd.read_csv(count_file, index_col=0)
    annot = pd.read_csv(annotation_file)

    # Clean dose column
    annot["dose"] = annot["dose"].astype(str)
    annot.loc[annot["dose"].isin(["nan", "NaN", "", "None"]), "dose"] = "0"

    # Flag controls
    annot["is_control"] = annot["compound"].isin(control_names)

    # Subset to this cell line: compound samples + control samples
    mask = (
        (annot["sample"] == cell_line)
        & ((annot["compound"] == compound_name) | annot["is_control"])
    )
    annot_sub = annot.loc[mask].copy()

    if len(annot_sub) < 2:
        print(f"ERROR: Not enough samples for {cell_line} {compound_name}")
        sys.exit(1)

    # Build condition column
    annot_sub["condition"] = np.where(annot_sub["is_control"], "control", annot_sub["dose"])

    # Subset count matrix to matching samples
    sample_names = annot_sub["sample_name"].tolist()
    counts_sub = counts[sample_names]

    return counts_sub, annot_sub





def run_pydeseq2(counts_sub, annot_sub):
    """
    Run pydeseq2: normalization + DEG analysis for each dose vs control.
    Returns normalized counts and DEG summary table.
    """
    # pydeseq2 expects samples as rows, genes as columns
    count_matrix = counts_sub.T
    count_matrix.index = annot_sub["sample_name"].values

    # Build metadata
    metadata = pd.DataFrame({
        "condition": annot_sub["condition"].values,
    }, index=annot_sub["sample_name"].values)

    # Create DESeq2 dataset and fit
    dds = DeseqDataSet(
        counts=count_matrix,
        metadata=metadata,
        design="~condition",
        refit_cooks=True,
        ref_level=["condition", "control"],
    )
    dds.deseq2()

    size_factors = pd.Series(dds.obs["size_factors"], index=dds.obs_names)
    norm_counts = counts_sub.div(size_factors[counts_sub.columns], axis=1)
    

    # ---- DEG results for each dose vs control ----
    conditions = [c for c in annot_sub["condition"].unique() if c != "control"]
    deg_frames = []

    for cond in conditions:
        stat = DeseqStats(dds, contrast=["condition", cond, "control"])
        stat.summary()
        res = stat.results_df.copy()
        res["gene"] = res.index
        res["dose"] = cond
        deg_frames.append(res)

    deg_all = pd.concat(deg_frames, ignore_index=True)

    # Standardize column names to match R output
    col_map = {
        "gene": "gene",
        "dose": "dose",
        "baseMean": "baseMean",
        "log2FoldChange": "log2FoldChange",
        "lfcSE": "lfcSE",
        "stat": "stat",
        "pvalue": "pvalue",
        "padj": "padj",
    }
    deg_all = deg_all.rename(columns=col_map)
    deg_all = deg_all[["gene", "dose", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj"]]

    return norm_counts, deg_all


def main():
    args = parse_args()

    # ---- Load and subset ----
    counts_sub, annot_sub = load_and_subset(
        args["count_file"],
        args["annotation_file"],
        args["cell_line"],
        args["compound_name"],
        args["control_names"],
    )


    # ---- Run normalization and DEG analysis ----
    norm_counts, deg_all = run_pydeseq2(counts_sub, annot_sub)

    # ---- Write outputs ----
    output_dir = Path(args["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)

    prefix = f"{args['cell_line']}_{args['compound_name']}"

    # Output 1: Normalized count matrix
    norm_out = output_dir / f"{prefix}_normalized_counts.csv"
    norm_df = norm_counts.copy()
    norm_df.insert(0, "gene", norm_df.index)
    norm_df.to_csv(norm_out, index=False)
    print(f"Written: {norm_out}")

    # Output 2: DEG summary table
    deg_out = output_dir / f"{prefix}_deg_summary.csv"
    deg_all.to_csv(deg_out, index=False)
    print(f"Written: {deg_out}")


    print("Done.")


if __name__ == "__main__":
    main()

File: pyBMD_modules/pre_pyBMD_normalize_filter.r
library("DESeq2")

args <- commandArgs(trailingOnly = TRUE)
count_file      <- args[1]
annotation_file <- args[2]
cell_line       <- args[3]
compound_name   <- args[4]
control_names   <- args[5]  # comma-separated, e.g. "DMSO1,DMSO2"
output_dir      <- args[6]

# ---- Parse control names ----
control_list <- trimws(unlist(strsplit(control_names, ",")))

# ---- Load data ----
cts <- as.matrix(read.csv(count_file, row.names = 1, check.names = FALSE))
annot <- read.csv(annotation_file, check.names = FALSE)

# ---- Clean dose column ----
annot$dose <- as.character(annot$dose)
annot$dose[is.na(annot$dose) | annot$dose == "NaN" | annot$dose == ""] <- "0"  # as some controls doses can be NaN or empty 

# ---- Flag controls by compound name ----
annot$is_control <- annot$compound %in% control_list

# ---- Subset to this cell line: compound + controls ----
annot_sub <- annot[
  annot$sample == cell_line &
  (annot$compound == compound_name | annot$is_control),
]

if (nrow(annot_sub) < 2) {
  stop(paste("Not enough samples for", cell_line, compound_name))
}

# ---- Build a clean condition label ----
annot_sub$condition <- ifelse(annot_sub$is_control, "control", annot_sub$dose)
annot_sub$condition <- factor(annot_sub$condition)

# ---- Build colData ----
colData <- data.frame(
  row.names = annot_sub$sample_name,
  condition = annot_sub$condition
)

# Subset and reorder count matrix
cts_sub <- cts[, rownames(colData)]


# ---- Run DESeq2 ----
dds <- DESeqDataSetFromMatrix(countData = cts_sub, colData = colData, design = ~ condition)
dds$condition <- relevel(dds$condition, ref = "control")
dds <- DESeq(dds)

# ============================================================
# OUTPUT 1: Normalized count matrix (genes x samples)
# ============================================================
norm_counts <- counts(dds, normalized = TRUE)
norm_counts_df <- as.data.frame(norm_counts)
norm_counts_df$gene <- rownames(norm_counts_df)
# Move gene column to front
norm_counts_df <- norm_counts_df[, c("gene", setdiff(names(norm_counts_df), "gene"))]

dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

norm_out <- file.path(output_dir, paste0(cell_line, "_", compound_name, "_normalized_counts.csv"))
write.csv(norm_counts_df, file = norm_out, row.names = FALSE)
cat("Written:", norm_out, "\n")

# ============================================================
# OUTPUT 2: DEG summary table (all doses combined)
# Used downstream to filter genes for BMD analysis (padj < 0.05)
# ============================================================
conditions <- levels(dds$condition)
conditions <- conditions[conditions != "control"]

deg_list <- list()

for (cond in conditions) {
  res <- results(dds, contrast = c("condition", cond, "control"))
  res_df <- as.data.frame(res)
  res_df$gene <- rownames(res_df)
  res_df$dose <- cond
  deg_list[[cond]] <- res_df
}

# Combine all dose results into one table
deg_all <- do.call(rbind, deg_list)
rownames(deg_all) <- NULL

# Reorder columns
deg_all <- deg_all[, c("gene", "dose", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj")]

deg_out <- file.path(output_dir, paste0(cell_line, "_", compound_name, "_deg_summary.csv"))
write.csv(deg_all, file = deg_out, row.names = FALSE)
cat("Written:", deg_out, "\n")



cat("Done.\n")